spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaDiscoveryClientConfiguration.java

```diff
@@ -51,52 +52,100 @@
-		private EurekaAutoServiceRegistration autoRegistration;
-
-		public void onApplicationEvent(RefreshScopeRefreshedEvent event) {
-			// This will force the creation of the EurekaClient bean if not already
-			// created
-			// to make sure the client will be re-registered after a refresh event
-			if (eurekaClient != null) {
-				eurekaClient.getApplications();
-			}
-			if (autoRegistration != null) {
-				// register in case meta data changed
-				this.autoRegistration.stop();
-				this.autoRegistration.start();
-			}
-		}
-
-	}
+    @Bean
+    @ConditionalOnMissingBean
+    public EurekaDiscoveryClient discoveryClient(EurekaClient client, EurekaClientConfig clientConfig) {
+        return new EurekaDiscoveryClient(client, clientConfig);
+    }
+
+    @Configuration(proxyBeanMethods = false)
+    @ConditionalOnProperty(value = "eureka.client.healthcheck.enabled", matchIfMissing = false)
+    protected static class EurekaHealthCheckHandlerConfiguration {
+
+        @Autowired(required = false)
+        private StatusAggregator statusAggregator = new SimpleStatusAggregator();
+
+        @Bean
+        @ConditionalOnMissingBean(HealthCheckHandler.class)
+        public EurekaHealthCheckHandler eurekaHealthCheckHandler() {
+            return new EurekaHealthCheckHandler(this.statusAggregator);
+        }
+
+    }
+
+    @Configuration(proxyBeanMethods = false)
+    @ConditionalOnClass(RefreshScopeRefreshedEvent.class)
+    protected static class EurekaClientConfigurationRefresher
+            implements ApplicationListener<RefreshScopeRefreshedEvent> {
+
+        private final Semaphore semaphore = new Semaphore(1);
+
+        @Autowired(required = false)
+        private EurekaClient eurekaClient;
+
+        @Autowired(required = false)
+        private EurekaAutoServiceRegistration autoRegistration;
+
+        public void onApplicationEvent(RefreshScopeRefreshedEvent event) {
+            try {
+                semaphore.acquire();
+                if (eurekaClient != null) {
+                    eurekaClient.getApplications();
+                }
+                if (autoRegistration != null) {
+                    // deregister instance
+                    this.autoRegistration.stop();
+                    // register instance
+                    this.autoRegistration.start();
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } finally {
+                semaphore.release();
+            }
+        }
+
+    }
+
+    @Configuration(proxyBeanMethods = false)
+    protected static class DiscoveryClientConfiguration {
+
+        private final Semaphore semaphore = new Semaphore(1);
+
+        @Autowired
+        private ApplicationInfoManager applicationInfoManager;
+
+        @Autowired
+        private HealthCheckHandler healthCheckHandler;
+
+        @Autowired
+        private InstanceInfo instanceInfo;
+
+        void refreshInstanceInfo() {
```

Comment:
It seems this method is not  ever used - please verify and finish up your changes, and then request another review.